broker/core/multiplexing/src/engine.cc

```diff
@@ -50,28 +52,46 @@ std::shared_ptr<engine> engine::instance_ptr() {
  * queue files.
  */
 void engine::load() {
-  SPDLOG_LOGGER_TRACE(log_v2::core(), "multiplexing: loading engine");
-  std::lock_guard<std::mutex> lk(_load_m);
+  auto logger = log_v2::instance().get(log_v2::CORE);
+  SPDLOG_LOGGER_TRACE(logger, "multiplexing: loading engine");
+  absl::MutexLock lk(&_load_m);
   if (!_instance)
-    _instance.reset(new engine);
+    _instance.reset(new engine(logger));
 }
 
 /**
- *  Unload class instance.
+ * @brief Unload the engine class instance.
  */
 void engine::unload() {
-  if (!_instance)
-    return;
-  if (_instance->_state != stopped)
-    _instance->stop();
+  SPDLOG_LOGGER_TRACE(log_v2::instance().get(log_v2::CORE),
+                      "multiplexing: unloading engine");
+  auto instance = instance_ptr();
+  if (instance) {
+    {
+      absl::ReleasableMutexLock lck(&instance->_kiew_m);
+      /* Here we wait for all the subscriber muxers to be stopped and removed
+       * from the muxers array. Even if they execute asynchronous functions,
+       * they have finished after that. */
+      auto muxers_empty = [&m = instance->_muxers,
+                           logger = instance->_logger]() {
+        logger->debug("Still {} muxers configured in Broker engine", m.size());
+        return m.empty();
+      };
+      instance->_logger->info("Waiting for the destruction of subscribers");
+      instance->_kiew_m.Await(absl::Condition(&muxers_empty));
+    }
 
-  SPDLOG_LOGGER_TRACE(log_v2::core(), "multiplexing: unloading engine");
-  std::lock_guard<std::mutex> lk(_load_m);
-  // Commit the cache file, if needed.
-  if (_instance && _instance->_cache_file)
-    _instance->_cache_file->commit();
+    absl::MutexLock lck(&_load_m);
+    instance->stop();
 
-  _instance.reset();
+    // Commit the cache file, if needed.
+    if (instance->_cache_file) {
+      // In case of muxers removed from the Engine and still events in _kiew
+      instance->publish(instance->_kiew);
```

Comment:
at this point _muxers is empty so publish and _send_to_subscribers do nothing?

Reply:
In publish(), we have the case `_state == stopped` where we put the events into the cache_file, that cache file that is an unprocessed file in this case.
Did I miss something?